[v2.0] Addresses page

nop33 opened the pull request.

nop33 commented on src/components/Ellipsed.tsx:
The changes in this file make use of a new simple hook I created (`useWindowResize`) and simplify the code massively. No changes in functionality.

nop33 commented on src/pages/UnlockedWallet/AddressesPage/AddressGridRow.tsx:
```diff
@@ -0,0 +1,203 @@
+
+  const assetsCellRef = useRef<HTMLDivElement>(null)
+
+  const ensureAssetsDisplayedInOnlyOneRow = useCallback(() => {
+    if (!assetsCellRef.current) return
+
+    if (assetsCellRef.current.offsetHeight > assetSymbolsRowHeight) {
+      const newArray = [...assetSymbolTexts]
+      const lastItem = newArray.pop()
+
+      if (lastItem?.startsWith('+')) {
+        const num = parseInt(lastItem.replace('+', ''))
+        newArray.pop()
+        newArray.push(`+${num + 1}`)
+      }
+      setAssetSymbolTexts(newArray)
+    } else if (!assetSymbolTexts.at(-1)?.startsWith('+') && assetSymbolTexts.length < assetsWithBalance.length) {
+      const newArray = [...assetSymbolTexts]
+      newArray.push(`+${assetsWithBalance.length - assetSymbolTexts.length}`)
+      setAssetSymbolTexts(newArray)
+    }
+  }, [assetSymbolTexts, assetsWithBalance.length])
+
+  useEffect(() => {
+    ensureAssetsDisplayedInOnlyOneRow()
+  }, [ensureAssetsDisplayedInOnlyOneRow])
+
+  useWindowResize(ensureAssetsDisplayedInOnlyOneRow)
```
This piece of code is responsible for this. It works when resizing the window as well, but only when scaling down 😅 

<img width="575" alt="image" src="https://user-images.githubusercontent.com/1579899/226979325-58aafea0-3774-4e9c-8f35-09c765cc972d.png">

nop33 replied:
it now works on all window resize! I wonder if you could give me a hand with framer motion so that we don't have that ugly glitch artifact.

nop33 replied:
Another idea I had in mind is to create an invisible clone of the `AssetSymbols` component instance and position it `absolute` so that it's over the visible one. First do the "add/remove" operation on the invisible component and after a short timeout, based on the content of the invisible container perform the same operation to the visible one. Hard to explain 😅

mvaivre replied:
Soooo... After playing a bit with this, I decided to drastically simplify it 🙃 

New design is also more consistent with other of our UIs (tx list, explorer). I really hope you don't mind!
We will probably be able to reuse your logic somewhere else.

nop33 commented on src/pages/UnlockedWallet/AddressesPage/AddressesTabContent.tsx:
```diff
@@ -95,3 +119,48 @@ const ToggleText = styled.div`
+`
+
+const TableGrid = styled(Box)`
+  contain: paint; // This is amazing. It replaces "overflow: hidden". Using "overflow" on this prevents us from having a sticky table header.
```
This took me many hours to figure out (look at the git history and you'll understand!)

mvaivre replied:
Oh wow! The number of times I was struggling with avoid the use of `overflow: hidden` to allow for "stickiness" 🙃.. <3

nop33 commented on src/components/SkeletonLoader.tsx:
The skeleton loaders were inspired by Notion's latest update 😃

mvaivre replied:
I really like it ✨ !

nop33 commented on public/electron.js:
```diff
@@ -151,7 +151,7 @@ function createWindow() {
     width: 1200,
     height: 800,
-    minWidth: 1000,
+    minWidth: 1200,
```
Making the wallet responsive with smaller than 1200 screens could be a nice task for a following release, WDYT?

Analytics from our users might help here...

mvaivre replied:
Ledger Live seems to be using a min width of 1020px so I decided to follow their lead, even though I found this to be a "lazy" strategy 🙃.
When our wallet will be used alongside dApps, it's gonna be useful to be able to reduce its size, so yes, working on a nice responsive layout should be considered for a future release. I'll create a task.

nop33 commented on src/pages/UnlockedWallet/AddressesPage/AddressGridRow.tsx:
```diff
@@ -59,17 +60,32 @@ const AddressGridRow = ({ addressHash, className }: AddressGridRowProps) => {
   const fiatBalance = calculateAmountWorth(BigInt(address.balance), price ?? 0)
 
+  const renderAssetLogos = () => {
+    const [displayedAssets, hiddenAssets] = chunk(assetsWithBalance, maxDisplayedAssets)
```
This is problematic because `assetsWithBalance` might contain in various places unknown tokens where the symbol is `undefined`. I'll update the code to fix it!

nop33 commented on src/pages/UnlockedWallet/AddressesPage/AddressGridRow.tsx:
```diff
@@ -59,17 +60,32 @@ const AddressGridRow = ({ addressHash, className }: AddressGridRowProps) => {
   const fiatBalance = calculateAmountWorth(BigInt(address.balance), price ?? 0)
 
+  const renderAssetLogos = () => {
+    const [displayedAssets, hiddenAssets] = chunk(assetsWithBalance, maxDisplayedAssets)
```
This way we are only taking the first 2 chunks and we're losing the rest! I'll update the code to fix that

mvaivre replied:
Arh, didn't notice that this fn is clever enough to create an infinite number of chunks if needed, and I only tested with a length of less than max * 2. Sorry about that.

mvaivre replied:
Haha I saw. I'm okay with that :)

mvaivre commented:
This looks better and better, delicious. 🍬 

Still a couple of issues though:

- After I click on "load more" a couple of times, the asset previously selected in the asset filter modal are deselected.
- The addresses filter seems to cumulate (linked to loading more TXs it seems):

<img width="713" alt="image" src="https://user-images.githubusercontent.com/6715943/228327038-4d9965c6-7823-4568-aa40-5197fcc7b8f7.png">

⬆️ I have 4 addresses, but it shows 8 selected.

- We should have a placeholder when no TXs correspond to a set of filter:

<img width="998" alt="image" src="https://user-images.githubusercontent.com/6715943/228327284-231da677-6544-4116-b058-50bbb27ebee0.png">

mvaivre reviewed (approved):
Pre approval! Code looks really good 👏  If you can take a look at the couple of issues I found, after that, feel free to merge 🚀

nop33 commented:
> The "copy address" shortcut on address hashes pops up above the action shortcuts :D

✅  Haha, what a nice catch! Fixed! 

> By messing around with the transfers page a lot (adding, removing filters, loading more TXs...) I end up at a point where I have 2 scroll contexts, and I can scroll the whole content (including the navigation sidebar) :

LOL! How the heck did you do that 😆 I've been playing forcefully with the transfers page, resizing the window, adding filters, removing, and scrolling, can't reproduce it though! I'll keep an eye on it, hopefully it appears! If it re-appears on your side, please do some element inspection to provide more feedback! It looks like the second scrollbar is the native scrollbar (not the one created by the `<Scrollbar>` component (which is only instantiated once on that page.

> Oof and that's a tricky one, I couldn't reproduce it so far. It happened as soon as I opened the Addresses page. Resizing the window made the glitch go away 👻

Oh, I have some ideas of what it is: It's probably because we changed our mono font so the calculation of the middle ellipsis is not working properly. Since the font is not native to the system and it needs to be loaded separately, it might cause some problems when calculating the character width. Or maybe it's a rare corner-case of the ellipsis mechanism even without the font change.

nop33 commented:
I'll merge even though I didn't manage to reproduce your last 2 comments. I created an issue for one of them.
